lib/src/command/add.dart

```diff
@@ -259,24 +307,33 @@ class AddCommand extends PubCommand {
     );
   }
 
-  /// Parse [package] to return the corresponding [PackageRange], as well as its
-  /// representation in `pubspec.yaml`.
+  /// Parse [package] to return the corresponding [_ParseResult].
   ///
   /// [package] must be written in the format
-  /// `<package-name>[:<version-constraint>]`, where quotations should be used
-  /// if necessary.
+  /// `dev:<package-name>[:descriptor>]`, where quotations should be used if
+  /// necessary.
+  ///
+  /// `descriptor` is what you would put in a pubspec.yaml in the dependencies
+  /// section.
   ///
   /// Examples:
   /// ```
   /// retry
   /// retry:2.0.0
-  /// retry:^2.0.0
+  /// dev:retry:^2.0.0
   /// retry:'>=2.0.0'
   /// retry:'>2.0.0 <3.0.1'
   /// 'retry:>2.0.0 <3.0.1'
   /// retry:any
+  /// 'retry:{"path":"../foo"}'
+  /// 'retry:{"git":{"url":"../foo","ref":"branchname"},"version":"^1.2.3"}'
+  /// 'retry:{"sdk":"flutter"}'
+  /// 'retry:{"hosted":"mypub.dev"}'
   /// ```
   ///
+  /// The --path --sdk and --git-<option> arguments cannot be combined with a
+  /// non-string descriptor.
```

Comment:
What about splitting this into two methods?
 * One for parsing when any of the options are given `_parsePackageAndConstraint` (which is deprecated)
 * One for parsing package and descriptor `_parsePackageAndDescriptor` (for the new syntax)

That way you can keep the old logic in `_parsePackageAndConstraint` and only add new logic in `_parsePackageAndDescriptor`, and you just look for options to decide which one to use.